diff --git a/mingw-w64-crt/Makefile.in b/mingw-w64-crt/Makefile.in
index 7512bfe42..74a487b51 100644
--- a/mingw-w64-crt/Makefile.in
+++ b/mingw-w64-crt/Makefile.in
@@ -11208,7 +11208,8 @@ src_msvcrt64 = \
   misc/_create_locale.c \
   misc/_free_locale.c \
   misc/_get_current_locale.c \
-  misc/seterrno.c
+  misc/seterrno.c \
+  misc/wassert.c
 
 
 # Files included in libmsvcrt-os.a (for msvcrt.dll) on arm32
diff --git a/mingw-w64-crt/lib-common/msvcrt.def.in b/mingw-w64-crt/lib-common/msvcrt.def.in
index 69068ae4b..5c6039720 100644
--- a/mingw-w64-crt/lib-common/msvcrt.def.in
+++ b/mingw-w64-crt/lib-common/msvcrt.def.in
@@ -1648,7 +1648,7 @@ _vwprintf_p_l
 _vwprintf_s_l
 ; _waccess_s replaced by emu
 ; _wasctime_s replaced by emu
-F_NON_I386(_wassert) ; i386 _wassert replaced by emu
+F_ARM_ANY(_wassert) ; i386 and x64 _wassert replaced by emu
 _wcscoll_l
 _wcsdup_dbg
 _wcserror_s
