diff --unified --recursive --text mingw-w64-v12.0.0.orig/mingw-w64-crt/lib-common/msvcrt.def.in mingw-w64-v12.0.0/mingw-w64-crt/lib-common/msvcrt.def.in
--- mingw-w64-v12.0.0.orig/mingw-w64-crt/lib-common/msvcrt.def.in
+++ mingw-w64-v12.0.0/mingw-w64-crt/lib-common/msvcrt.def.in
@@ -1648,7 +1648,7 @@
 _vwprintf_s_l
 ; _waccess_s replaced by emu
 ; _wasctime_s replaced by emu
-F_NON_I386(_wassert) ; i386 _wassert replaced by emu
+F_ARM_ANY(_wassert) ; i386 and x64 _wassert replaced by emu
 _wcscoll_l
 _wcsdup_dbg
 _wcserror_s
diff --unified --recursive --text mingw-w64-v12.0.0.orig/mingw-w64-crt/Makefile.am mingw-w64-v12.0.0/mingw-w64-crt/Makefile.am
--- mingw-w64-v12.0.0.orig/mingw-w64-crt/Makefile.am
+++ mingw-w64-v12.0.0/mingw-w64-crt/Makefile.am
@@ -460,7 +460,8 @@
   misc/_create_locale.c \
   misc/_free_locale.c \
   misc/_get_current_locale.c \
-  misc/seterrno.c
+  misc/seterrno.c \
+  misc/wassert.c
 
 # Files included in libmsvcrt-os.a (for msvcrt.dll) on arm32
 src_msvcrtarm32=\
diff --unified --recursive --text mingw-w64-v12.0.0.orig/mingw-w64-crt/Makefile.in mingw-w64-v12.0.0/mingw-w64-crt/Makefile.in
--- mingw-w64-v12.0.0.orig/mingw-w64-crt/Makefile.in
+++ mingw-w64-v12.0.0/mingw-w64-crt/Makefile.in
@@ -2967,7 +2967,8 @@
 	secapi/wmemmove_s.c stdio/fseeki64.c stdio/mingw_lock.c \
 	misc/__p___argv.c misc/__p__acmdln.c misc/__p__commode.c \
 	misc/__p__fmode.c misc/__p__wcmdln.c misc/_create_locale.c \
-	misc/_free_locale.c misc/_get_current_locale.c misc/seterrno.c
+	misc/_free_locale.c misc/_get_current_locale.c misc/seterrno.c \
+	misc/wassert.c
 am__objects_96 =  \
 	misc/lib64_libmsvcrt_extra_a-_configthreadlocale.$(OBJEXT) \
 	misc/lib64_libmsvcrt_extra_a-imaxdiv.$(OBJEXT) \
@@ -3028,7 +3029,8 @@
 	misc/lib64_libmsvcrt_extra_a-_create_locale.$(OBJEXT) \
 	misc/lib64_libmsvcrt_extra_a-_free_locale.$(OBJEXT) \
 	misc/lib64_libmsvcrt_extra_a-_get_current_locale.$(OBJEXT) \
-	misc/lib64_libmsvcrt_extra_a-seterrno.$(OBJEXT)
+	misc/lib64_libmsvcrt_extra_a-seterrno.$(OBJEXT) \
+	misc/lib64_libmsvcrt_extra_a-wassert.$(OBJEXT)
 @LIB64_TRUE@@W32API_FALSE@am_lib64_libmsvcrt_extra_a_OBJECTS =  \
 @LIB64_TRUE@@W32API_FALSE@	$(am__objects_97)
 lib64_libmsvcrt_extra_a_OBJECTS =  \
@@ -8804,6 +8806,7 @@
 	misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-output_format.Po \
 	misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-purecall.Po \
 	misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-seterrno.Po \
+	misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Po \
 	misc/$(DEPDIR)/lib64_libucrt_extra_a-__initenv.Po \
 	misc/$(DEPDIR)/lib64_libucrt_extra_a-ucrt-access.Po \
 	misc/$(DEPDIR)/lib64_libucrtapp_extra_a-longjmp.Po \
@@ -11208,7 +11211,8 @@
   misc/_create_locale.c \
   misc/_free_locale.c \
   misc/_get_current_locale.c \
-  misc/seterrno.c
+  misc/seterrno.c \
+  misc/wassert.c
 
 
 # Files included in libmsvcrt-os.a (for msvcrt.dll) on arm32
@@ -17739,6 +17743,8 @@
 	misc/$(am__dirstamp) misc/$(DEPDIR)/$(am__dirstamp)
 misc/lib64_libmsvcrt_extra_a-seterrno.$(OBJEXT): misc/$(am__dirstamp) \
 	misc/$(DEPDIR)/$(am__dirstamp)
+misc/lib64_libmsvcrt_extra_a-wassert.$(OBJEXT): misc/$(am__dirstamp) \
+	misc/$(DEPDIR)/$(am__dirstamp)
 
 lib64/libmsvcrt_extra.a: $(lib64_libmsvcrt_extra_a_OBJECTS) $(lib64_libmsvcrt_extra_a_DEPENDENCIES) $(EXTRA_lib64_libmsvcrt_extra_a_DEPENDENCIES) lib64/$(am__dirstamp)
 	$(AM_V_at)-rm -f lib64/libmsvcrt_extra.a
@@ -24959,6 +24965,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-output_format.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-purecall.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-seterrno.Po@am__quote@ # am--include-marker
+@AMDEP_TRUE@@am__include@ @am__quote@misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@misc/$(DEPDIR)/lib64_libucrt_extra_a-__initenv.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@misc/$(DEPDIR)/lib64_libucrt_extra_a-ucrt-access.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@misc/$(DEPDIR)/lib64_libucrtapp_extra_a-longjmp.Po@am__quote@ # am--include-marker
@@ -48139,6 +48146,20 @@
 @AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
 @am__fastdepCC_FALSE@	$(AM_V_CC@am__nodep@)$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(lib64_libmsvcrt_extra_a_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o misc/lib64_libmsvcrt_extra_a-seterrno.obj `if test -f 'misc/seterrno.c'; then $(CYGPATH_W) 'misc/seterrno.c'; else $(CYGPATH_W) '$(srcdir)/misc/seterrno.c'; fi`
 
+misc/lib64_libmsvcrt_extra_a-wassert.o: misc/wassert.c
+@am__fastdepCC_TRUE@	$(AM_V_CC)$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(lib64_libmsvcrt_extra_a_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -MT misc/lib64_libmsvcrt_extra_a-wassert.o -MD -MP -MF misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Tpo -c -o misc/lib64_libmsvcrt_extra_a-wassert.o `test -f 'misc/wassert.c' || echo '$(srcdir)/'`misc/wassert.c
+@am__fastdepCC_TRUE@	$(AM_V_at)$(am__mv) misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Tpo misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Po
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	$(AM_V_CC)source='misc/wassert.c' object='misc/lib64_libmsvcrt_extra_a-wassert.o' libtool=no @AMDEPBACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
+@am__fastdepCC_FALSE@	$(AM_V_CC@am__nodep@)$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(lib64_libmsvcrt_extra_a_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o misc/lib64_libmsvcrt_extra_a-wassert.o `test -f 'misc/wassert.c' || echo '$(srcdir)/'`misc/wassert.c
+
+misc/lib64_libmsvcrt_extra_a-wassert.obj: misc/wassert.c
+@am__fastdepCC_TRUE@	$(AM_V_CC)$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(lib64_libmsvcrt_extra_a_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -MT misc/lib64_libmsvcrt_extra_a-wassert.obj -MD -MP -MF misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Tpo -c -o misc/lib64_libmsvcrt_extra_a-wassert.obj `if test -f 'misc/wassert.c'; then $(CYGPATH_W) 'misc/wassert.c'; else $(CYGPATH_W) '$(srcdir)/misc/wassert.c'; fi`
+@am__fastdepCC_TRUE@	$(AM_V_at)$(am__mv) misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Tpo misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Po
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	$(AM_V_CC)source='misc/wassert.c' object='misc/lib64_libmsvcrt_extra_a-wassert.obj' libtool=no @AMDEPBACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
+@am__fastdepCC_FALSE@	$(AM_V_CC@am__nodep@)$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(lib64_libmsvcrt_extra_a_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o misc/lib64_libmsvcrt_extra_a-wassert.obj `if test -f 'misc/wassert.c'; then $(CYGPATH_W) 'misc/wassert.c'; else $(CYGPATH_W) '$(srcdir)/misc/wassert.c'; fi`
+
 libsrc/lib64_libmsxml2_a-msxml2.o: libsrc/msxml2.c
 @am__fastdepCC_TRUE@	$(AM_V_CC)$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(lib64_libmsxml2_a_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -MT libsrc/lib64_libmsxml2_a-msxml2.o -MD -MP -MF libsrc/$(DEPDIR)/lib64_libmsxml2_a-msxml2.Tpo -c -o libsrc/lib64_libmsxml2_a-msxml2.o `test -f 'libsrc/msxml2.c' || echo '$(srcdir)/'`libsrc/msxml2.c
 @am__fastdepCC_TRUE@	$(AM_V_at)$(am__mv) libsrc/$(DEPDIR)/lib64_libmsxml2_a-msxml2.Tpo libsrc/$(DEPDIR)/lib64_libmsxml2_a-msxml2.Po
@@ -76365,6 +76386,7 @@
 	-rm -f misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-output_format.Po
 	-rm -f misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-purecall.Po
 	-rm -f misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-seterrno.Po
+	-rm -f misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Po
 	-rm -f misc/$(DEPDIR)/lib64_libucrt_extra_a-__initenv.Po
 	-rm -f misc/$(DEPDIR)/lib64_libucrt_extra_a-ucrt-access.Po
 	-rm -f misc/$(DEPDIR)/lib64_libucrtapp_extra_a-longjmp.Po
@@ -79783,6 +79805,7 @@
 	-rm -f misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-output_format.Po
 	-rm -f misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-purecall.Po
 	-rm -f misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-seterrno.Po
+	-rm -f misc/$(DEPDIR)/lib64_libmsvcrt_extra_a-wassert.Po
 	-rm -f misc/$(DEPDIR)/lib64_libucrt_extra_a-__initenv.Po
 	-rm -f misc/$(DEPDIR)/lib64_libucrt_extra_a-ucrt-access.Po
 	-rm -f misc/$(DEPDIR)/lib64_libucrtapp_extra_a-longjmp.Po
