diff --unified --recursive --text gcc-14.1.0.orig/libstdc++-v3/config.h.in gcc-14.1.0/libstdc++-v3/config.h.in
--- gcc-14.1.0.orig/libstdc++-v3/config.h.in
+++ gcc-14.1.0/libstdc++-v3/config.h.in
@@ -541,9 +541,6 @@
 /* Define to 1 if you have the <xlocale.h> header file. */
 #undef HAVE_XLOCALE_H
 
-/* Define to 1 if you have the `_aligned_malloc' function. */
-#undef HAVE__ALIGNED_MALLOC
-
 /* Define to 1 if you have the `_wfopen' function. */
 #undef HAVE__WFOPEN
 
