diff --unified --recursive --text gcc-13.2.0.orig/libstdc++-v3/config.h.in gcc-13.2.0/libstdc++-v3/config.h.in
--- gcc-13.2.0.orig/libstdc++-v3/config.h.in
+++ gcc-13.2.0/libstdc++-v3/config.h.in
@@ -540,9 +540,6 @@
 /* Define to 1 if you have the `_acosl' function. */
 #undef HAVE__ACOSL
 
-/* Define to 1 if you have the `_aligned_malloc' function. */
-#undef HAVE__ALIGNED_MALLOC
-
 /* Define to 1 if you have the `_asinf' function. */
 #undef HAVE__ASINF
 
