diff --unified --recursive --text gdb-14.2.orig/gdb/ser-tcp.c gdb-14.2/gdb/ser-tcp.c
--- gdb-14.2.orig/gdb/ser-tcp.c
+++ gdb-14.2/gdb/ser-tcp.c
@@ -39,7 +39,7 @@
 #include "gdbsupport/gdb_sys_time.h"
 
 #ifdef USE_WIN32API
-#include <ws2tcpip.h>
+#include "win32-thunk.h"
 #ifndef ETIMEDOUT
 #define ETIMEDOUT WSAETIMEDOUT
 #endif
@@ -290,7 +290,7 @@
   if (parsed.port_str.empty ())
     error (_("Missing port on hostname '%s'"), name);
 
-  int r = getaddrinfo (parsed.host_str.c_str (),
+  int r = Win32Thunk::getaddrinfo (parsed.host_str.c_str (),
 		       parsed.port_str.c_str (),
 		       &hint, &ainfo);
 
diff --unified --recursive --text gdb-14.2.orig/gdbserver/gdbreplay.cc gdb-14.2/gdbserver/gdbreplay.cc
--- gdb-14.2.orig/gdbserver/gdbreplay.cc
+++ gdb-14.2/gdbserver/gdbreplay.cc
@@ -53,7 +53,7 @@
 #endif
 
 #if USE_WIN32API
-#include <ws2tcpip.h>
+#include "win32-thunk.h"
 #endif
 
 #include "gdbsupport/netstuff.h"
@@ -154,7 +154,7 @@
     }
 #endif
 
-  int r = getaddrinfo (parsed.host_str.c_str (), parsed.port_str.c_str (),
+  int r = Win32Thunk::getaddrinfo (parsed.host_str.c_str (), parsed.port_str.c_str (),
 		       &hint, &ainfo);
 
   if (r != 0)
@@ -230,7 +230,7 @@
       setsockopt (remote_desc_in, IPPROTO_TCP, TCP_NODELAY,
 		  (char *) &tmp, sizeof (tmp));
 
-      if (getnameinfo ((struct sockaddr *) &sockaddr, sockaddrsize,
+      if (Win32Thunk::getnameinfo ((struct sockaddr *) &sockaddr, sockaddrsize,
 		       orig_host, sizeof (orig_host),
 		       orig_port, sizeof (orig_port),
 		       NI_NUMERICHOST | NI_NUMERICSERV) == 0)
diff --unified --recursive --text gdb-14.2.orig/gdbserver/remote-utils.cc gdb-14.2/gdbserver/remote-utils.cc
--- gdb-14.2.orig/gdbserver/remote-utils.cc
+++ gdb-14.2/gdbserver/remote-utils.cc
@@ -65,7 +65,7 @@
 #include <sys/stat.h>
 
 #if USE_WIN32API
-#include <ws2tcpip.h>
+#include "win32-thunk.h"
 #endif
 
 #ifndef HAVE_SOCKLEN_T
@@ -182,7 +182,7 @@
   /* Convert IP address to string.  */
   char orig_host[GDB_NI_MAX_ADDR], orig_port[GDB_NI_MAX_PORT];
 
-  int r = getnameinfo ((struct sockaddr *) &sockaddr, len,
+  int r = Win32Thunk::getnameinfo ((struct sockaddr *) &sockaddr, len,
 		       orig_host, sizeof (orig_host),
 		       orig_port, sizeof (orig_port),
 		       NI_NUMERICHOST | NI_NUMERICSERV);
@@ -260,7 +260,7 @@
     }
 #endif
 
-  int r = getaddrinfo (parsed.host_str.c_str (), parsed.port_str.c_str (),
+  int r = Win32Thunk::getaddrinfo (parsed.host_str.c_str (), parsed.port_str.c_str (),
 		       &hint, &ainfo);
 
   if (r != 0)
@@ -387,7 +387,7 @@
       if (getsockname (listen_desc, (struct sockaddr *) &sockaddr, &len) < 0)
 	perror_with_name ("Can't determine port");
 
-      int r = getnameinfo ((struct sockaddr *) &sockaddr, len,
+      int r = Win32Thunk::getnameinfo ((struct sockaddr *) &sockaddr, len,
 			   NULL, 0,
 			   listen_port, sizeof (listen_port),
 			   NI_NUMERICSERV);
diff --unified --recursive --text gdb-14.2.orig/gdbsupport/netstuff.cc gdb-14.2/gdbsupport/netstuff.cc
--- gdb-14.2.orig/gdbsupport/netstuff.cc
+++ gdb-14.2/gdbsupport/netstuff.cc
@@ -21,7 +21,7 @@
 #include <algorithm>
 
 #ifdef USE_WIN32API
-#include <ws2tcpip.h>
+#include "win32-thunk.h"
 #else
 #include <netinet/in.h>
 #include <arpa/inet.h>
@@ -34,7 +34,7 @@
 
 scoped_free_addrinfo::~scoped_free_addrinfo ()
 {
-  freeaddrinfo (m_res);
+  Win32Thunk::freeaddrinfo (m_res);
 }
 
 /* See gdbsupport/netstuff.h.  */
