diff --git a/gdb/ser-tcp.c b/gdb/ser-tcp.c
index 019aa4e52ec..f71684785f0 100644
--- a/gdb/ser-tcp.c
+++ b/gdb/ser-tcp.c
@@ -39,7 +39,7 @@
 #include "gdbsupport/gdb_sys_time.h"
 
 #ifdef USE_WIN32API
-#include <ws2tcpip.h>
+#include "win32-thunk.h"
 #ifndef ETIMEDOUT
 #define ETIMEDOUT WSAETIMEDOUT
 #endif
@@ -290,7 +290,7 @@ net_open (struct serial *scb, const char *name)
   if (parsed.port_str.empty ())
     error (_("Missing port on hostname '%s'"), name);
 
-  int r = getaddrinfo (parsed.host_str.c_str (),
+  int r = Win32Thunk::getaddrinfo (parsed.host_str.c_str (),
 		       parsed.port_str.c_str (),
 		       &hint, &ainfo);
 
diff --git a/gdbserver/gdbreplay.cc b/gdbserver/gdbreplay.cc
index 438a6a60bdb..5272533734e 100644
--- a/gdbserver/gdbreplay.cc
+++ b/gdbserver/gdbreplay.cc
@@ -53,7 +53,7 @@
 #endif
 
 #if USE_WIN32API
-#include <ws2tcpip.h>
+#include "win32-thunk.h"
 #endif
 
 #include "gdbsupport/netstuff.h"
@@ -192,7 +192,7 @@ remote_open (char *name)
     }
 #endif
 
-  int r = getaddrinfo (parsed.host_str.c_str (), parsed.port_str.c_str (),
+  int r = Win32Thunk::getaddrinfo (parsed.host_str.c_str (), parsed.port_str.c_str (),
 		       &hint, &ainfo);
 
   if (r != 0)
@@ -268,7 +268,7 @@ remote_open (char *name)
       setsockopt (remote_desc, IPPROTO_TCP, TCP_NODELAY,
 		  (char *) &tmp, sizeof (tmp));
 
-      if (getnameinfo ((struct sockaddr *) &sockaddr, sockaddrsize,
+      if (Win32Thunk::getnameinfo ((struct sockaddr *) &sockaddr, sockaddrsize,
 		       orig_host, sizeof (orig_host),
 		       orig_port, sizeof (orig_port),
 		       NI_NUMERICHOST | NI_NUMERICSERV) == 0)
diff --git a/gdbserver/remote-utils.cc b/gdbserver/remote-utils.cc
index 1195ce87ec7..0d102429a24 100644
--- a/gdbserver/remote-utils.cc
+++ b/gdbserver/remote-utils.cc
@@ -65,7 +65,7 @@
 #include <sys/stat.h>
 
 #if USE_WIN32API
-#include <ws2tcpip.h>
+#include "win32-thunk.h"
 #endif
 
 #ifndef HAVE_SOCKLEN_T
@@ -179,7 +179,7 @@ handle_accept_event (int err, gdb_client_data client_data)
   /* Convert IP address to string.  */
   char orig_host[GDB_NI_MAX_ADDR], orig_port[GDB_NI_MAX_PORT];
 
-  int r = getnameinfo ((struct sockaddr *) &sockaddr, len,
+  int r = Win32Thunk::getnameinfo ((struct sockaddr *) &sockaddr, len,
 		       orig_host, sizeof (orig_host),
 		       orig_port, sizeof (orig_port),
 		       NI_NUMERICHOST | NI_NUMERICSERV);
@@ -257,7 +257,7 @@ remote_prepare (const char *name)
     }
 #endif
 
-  int r = getaddrinfo (parsed.host_str.c_str (), parsed.port_str.c_str (),
+  int r = Win32Thunk::getaddrinfo (parsed.host_str.c_str (), parsed.port_str.c_str (),
 		       &hint, &ainfo);
 
   if (r != 0)
@@ -384,7 +384,7 @@ remote_open (const char *name)
       if (getsockname (listen_desc, (struct sockaddr *) &sockaddr, &len) < 0)
 	perror_with_name ("Can't determine port");
 
-      int r = getnameinfo ((struct sockaddr *) &sockaddr, len,
+      int r = Win32Thunk::getnameinfo ((struct sockaddr *) &sockaddr, len,
 			   NULL, 0,
 			   listen_port, sizeof (listen_port),
 			   NI_NUMERICSERV);
diff --git a/gdbsupport/netstuff.cc b/gdbsupport/netstuff.cc
index a056e6adf7d..246239ffcbd 100644
--- a/gdbsupport/netstuff.cc
+++ b/gdbsupport/netstuff.cc
@@ -21,7 +21,7 @@
 #include <algorithm>
 
 #ifdef USE_WIN32API
-#include <ws2tcpip.h>
+#include "win32-thunk.h"
 #else
 #include <netinet/in.h>
 #include <arpa/inet.h>
@@ -34,7 +34,7 @@
 
 scoped_free_addrinfo::~scoped_free_addrinfo ()
 {
-  freeaddrinfo (m_res);
+  Win32Thunk::freeaddrinfo (m_res);
 }
 
 /* See gdbsupport/netstuff.h.  */
