diff --unified --recursive --text gdb-14.2.orig/gdb/mingw-hdep.c gdb-14.2/gdb/mingw-hdep.c
--- gdb-14.2.orig/gdb/mingw-hdep.c
+++ gdb-14.2/gdb/mingw-hdep.c
@@ -24,7 +24,7 @@
 #include "gdbsupport/gdb_select.h"
 #include "inferior.h"
 
-#include <windows.h>
+#include "win32-thunk.h"
 #include <signal.h>
 
 /* Return an absolute file name of the running GDB, if possible, or
@@ -382,7 +382,7 @@
   DWORD len = 0;
   while (true)
     {
-      len = GetConsoleProcessList (results.data (), results.size ());
+      len = Win32Thunk::GetConsoleProcessList (results.data (), results.size ());
       /* Note that LEN == 0 is a failure, but we can treat it the same
 	 as a "no".  */
       if (len <= results.size ())
diff --unified --recursive --text gdb-14.2.orig/gdb/nat/windows-nat.c gdb-14.2/gdb/nat/windows-nat.c
--- gdb-14.2.orig/gdb/nat/windows-nat.c
+++ gdb-14.2/gdb/nat/windows-nat.c
@@ -572,7 +572,7 @@
     {
       /* This fails on 32bit Windows because it has no SysWOW64 directory,
 	 and in this case a path conversion isn't necessary.  */
-      UINT len = GetSystemWow64DirectoryA (syswow_dir, sizeof (syswow_dir));
+      UINT len = Win32Thunk::GetSystemWow64DirectoryA (syswow_dir, sizeof (syswow_dir));
       if (len > 0)
 	{
 	  /* Check that we have passed a large enough buffer.  */
diff --unified --recursive --text gdb-14.2.orig/gdb/nat/windows-nat.h gdb-14.2/gdb/nat/windows-nat.h
--- gdb-14.2.orig/gdb/nat/windows-nat.h
+++ gdb-14.2/gdb/nat/windows-nat.h
@@ -19,7 +19,7 @@
 #ifndef NAT_WINDOWS_NAT_H
 #define NAT_WINDOWS_NAT_H
 
-#include <windows.h>
+#include "win32-thunk.h"
 #include <psapi.h>
 #include <vector>
 
