diff --git a/gdb/nat/windows-nat.h b/gdb/nat/windows-nat.h
index 84f6505e4c7..aaeedcc8ac1 100644
--- a/gdb/nat/windows-nat.h
+++ b/gdb/nat/windows-nat.h
@@ -19,7 +19,7 @@
 #ifndef NAT_WINDOWS_NAT_H
 #define NAT_WINDOWS_NAT_H
 
-#include <windows.h>
+#include "win32-thunk.h"
 #include <vector>
 
 #include "gdbsupport/gdb_optional.h"
diff --git a/gdb/windows-nat.c b/gdb/windows-nat.c
index 108e1e32edc..05c25c95587 100644
--- a/gdb/windows-nat.c
+++ b/gdb/windows-nat.c
@@ -1912,7 +1912,7 @@ windows_add_all_dlls (void)
     {
       /* This fails on 32bit Windows because it has no SysWOW64 directory,
 	 and in this case a path conversion isn't necessary.  */
-      UINT len = GetSystemWow64DirectoryA (syswow_dir, sizeof (syswow_dir));
+      UINT len = Win32Thunk::GetSystemWow64DirectoryA (syswow_dir, sizeof (syswow_dir));
       if (len > 0)
 	{
 	  /* Check that we have passed a large enough buffer.  */
