diff --git a/gdb/nat/windows-nat.c b/gdb/nat/windows-nat.c
index 27d7911a30e..8577f37a108 100644
--- a/gdb/nat/windows-nat.c
+++ b/gdb/nat/windows-nat.c
@@ -404,7 +404,7 @@ windows_add_dll (LPVOID load_addr)
     {
       /* This fails on 32bit Windows because it has no SysWOW64 directory,
 	 and in this case a path conversion isn't necessary.  */
-      UINT len = GetSystemWow64DirectoryA (syswow_dir, sizeof (syswow_dir));
+      UINT len = Win32Thunk::GetSystemWow64DirectoryA (syswow_dir, sizeof (syswow_dir));
       if (len > 0)
 	{
 	  /* Check that we have passed a large enough buffer.  */
diff --git a/gdb/nat/windows-nat.h b/gdb/nat/windows-nat.h
index f0abd7d795c..5d135ad947b 100644
--- a/gdb/nat/windows-nat.h
+++ b/gdb/nat/windows-nat.h
@@ -19,7 +19,7 @@
 #ifndef NAT_WINDOWS_NAT_H
 #define NAT_WINDOWS_NAT_H
 
-#include <windows.h>
+#include "win32-thunk.h"
 #include <psapi.h>
 #include <vector>
 
