diff --git a/bfd/config.in b/bfd/config.in
index 7458362922c..5e2e34c63b5 100644
--- a/bfd/config.in
+++ b/bfd/config.in
@@ -366,9 +366,6 @@
 # define _DARWIN_USE_64_BIT_INODE 1
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* Define for large files, on AIX-style hosts. */
 #undef _LARGE_FILES
 
diff --git a/gdb/config.in b/gdb/config.in
index 0c144c8918b..964bda6891c 100644
--- a/gdb/config.in
+++ b/gdb/config.in
@@ -817,9 +817,6 @@
 # define _DARWIN_USE_64_BIT_INODE 1
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* Define for large files, on AIX-style hosts. */
 #undef _LARGE_FILES
 
diff --git a/gdbserver/config.in b/gdbserver/config.in
index 47e1e722fe0..fd6022490b4 100644
--- a/gdbserver/config.in
+++ b/gdbserver/config.in
@@ -506,9 +506,6 @@
 # define _DARWIN_USE_64_BIT_INODE 1
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* Define for large files, on AIX-style hosts. */
 #undef _LARGE_FILES
 
diff --git a/gdbsupport/config.in b/gdbsupport/config.in
index 832c92b2cb1..f7800385a3d 100644
--- a/gdbsupport/config.in
+++ b/gdbsupport/config.in
@@ -420,9 +420,6 @@
 # define _DARWIN_USE_64_BIT_INODE 1
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* Define for large files, on AIX-style hosts. */
 #undef _LARGE_FILES
 
diff --git a/gnulib/config.in b/gnulib/config.in
index f70c6a52b43..ed07fa679d3 100644
--- a/gnulib/config.in
+++ b/gnulib/config.in
@@ -1471,9 +1471,6 @@
 # endif
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* True if the compiler says it groks GNU C version MAJOR.MINOR.  */
 #if defined __GNUC__ && defined __GNUC_MINOR__
 # define _GL_GNUC_PREREQ(major, minor) \
@@ -1533,9 +1530,6 @@
 /* For standard stat data types on VMS. */
 #undef _USE_STD_STAT
 
-/* For 64-bit time_t on 32-bit mingw. */
-#undef __MINGW_USE_VC2005_COMPAT
-
 /* Define to 1 if the system <stdint.h> predates C++11. */
 #undef __STDC_CONSTANT_MACROS
 
diff --git a/libbacktrace/config.h.in b/libbacktrace/config.h.in
index ee2616335c7..e2ac2c89fcb 100644
--- a/libbacktrace/config.h.in
+++ b/libbacktrace/config.h.in
@@ -180,9 +180,6 @@
 # define _DARWIN_USE_64_BIT_INODE 1
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* Define for large files, on AIX-style hosts. */
 #undef _LARGE_FILES
 
diff --git a/libctf/config.h.in b/libctf/config.h.in
index 794779eab5a..7d8c91cf34b 100644
--- a/libctf/config.h.in
+++ b/libctf/config.h.in
@@ -185,9 +185,6 @@
 # define _DARWIN_USE_64_BIT_INODE 1
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* Define for large files, on AIX-style hosts. */
 #undef _LARGE_FILES
 
diff --git a/libiberty/config.in b/libiberty/config.in
index 1b1f2b09a8a..4c3835e815e 100644
--- a/libiberty/config.in
+++ b/libiberty/config.in
@@ -546,9 +546,6 @@
 # define _DARWIN_USE_64_BIT_INODE 1
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* Define for large files, on AIX-style hosts. */
 #undef _LARGE_FILES
 
diff --git a/libsframe/config.h.in b/libsframe/config.h.in
index 74810fd30df..69b745feca1 100644
--- a/libsframe/config.h.in
+++ b/libsframe/config.h.in
@@ -115,9 +115,6 @@
 # define _DARWIN_USE_64_BIT_INODE 1
 #endif
 
-/* Number of bits in a file offset, on hosts where this is settable. */
-#undef _FILE_OFFSET_BITS
-
 /* Define for large files, on AIX-style hosts. */
 #undef _LARGE_FILES
 
diff --git a/readline/readline/config.h.in b/readline/readline/config.h.in
index 6de47b58743..bd448ded30b 100644
--- a/readline/readline/config.h.in
+++ b/readline/readline/config.h.in
@@ -14,8 +14,6 @@
    characters, even if the OS supports them. */
 #undef NO_MULTIBYTE_SUPPORT
 
-#undef _FILE_OFFSET_BITS
-
 /* Define as the return type of signal handlers (int or void).  */
 #undef RETSIGTYPE
 
