diff --git a/Include/internal/pycore_condvar.h b/Include/internal/pycore_condvar.h
index 2177bed31fc..7ad87da9b02 100644
--- a/Include/internal/pycore_condvar.h
+++ b/Include/internal/pycore_condvar.h
@@ -5,6 +5,10 @@
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+#ifdef NT_THREADS
+#  undef _POSIX_THREADS
+#endif
+
 #ifndef _POSIX_THREADS
 /* This means pthreads are not implemented in libc headers, hence the macro
    not present in unistd.h. But they still can be implemented as an external
@@ -41,6 +45,7 @@
 #  define WIN32_LEAN_AND_MEAN
 #endif
 #include <windows.h>              // CRITICAL_SECTION
+#undef _POSIX_THREADS
 
 /* options */
 /* non-emulated condition variables are provided for those that want
diff --git a/Include/pythread.h b/Include/pythread.h
index 63714437c49..42bd2adfe44 100644
--- a/Include/pythread.h
+++ b/Include/pythread.h
@@ -7,6 +7,10 @@ typedef void *PyThread_type_lock;
 extern "C" {
 #endif
 
+#ifdef NT_THREADS
+#  undef _POSIX_THREADS
+#endif
+
 /* Return status codes for Python lock acquisition.  Chosen for maximum
  * backwards compatibility, ie failure -> 0, success -> 1.  */
 typedef enum PyLockStatus {
diff --git a/Modules/_winapi.c b/Modules/_winapi.c
index ac3ee113ffd..c28c01db120 100644
--- a/Modules/_winapi.c
+++ b/Modules/_winapi.c
@@ -571,7 +571,7 @@ _winapi_CreateJunction_impl(PyObject *module, LPCWSTR src_path,
     tp.base.PrivilegeCount = 1;
     tp.base.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;
     if (!AdjustTokenPrivileges(token, FALSE, &tp.base, sizeof(previousTp),
-                               &previousTp.base, &previousTpSize)) {
+                               &previousTp.base, (DWORD *)&previousTpSize)) {
         goto cleanup;
     }
 
diff --git a/Modules/posixmodule.c b/Modules/posixmodule.c
index c21c6f06c72..a8f388706d0 100644
--- a/Modules/posixmodule.c
+++ b/Modules/posixmodule.c
@@ -39,6 +39,7 @@
 #  if defined(MS_WINDOWS_DESKTOP) || defined(MS_WINDOWS_SYSTEM)
 #    define HAVE_SYMLINK
 #  endif /* MS_WINDOWS_DESKTOP | MS_WINDOWS_SYSTEM */
+#  include <process.h>
 #endif
 
 #include "structmember.h"         // PyMemberDef
@@ -4961,7 +4962,7 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
     }
 
     Py_BEGIN_ALLOW_THREADS
-    ret = PathCchSkipRoot(buffer, &end);
+    ret = PathCchSkipRoot(buffer, (const wchar_t **)&end);
     Py_END_ALLOW_THREADS
     if (FAILED(ret)) {
         result = Py_BuildValue("sO", "", path->object);
diff --git a/Python/sysmodule.c b/Python/sysmodule.c
index 6ed28ec5fbd..ea4c09155dc 100644
--- a/Python/sysmodule.c
+++ b/Python/sysmodule.c
@@ -3432,6 +3432,8 @@ _PySys_InitCore(PyThreadState *tstate, PyObject *sysdict)
 #ifdef MS_COREDLL
     SET_SYS("dllhandle", PyLong_FromVoidPtr(PyWin_DLLhModule));
     SET_SYS_FROM_STRING("winver", PyWin_DLLVersionString);
+#else
+    SET_SYS_FROM_STRING("winver", "");
 #endif
 #ifdef ABIFLAGS
     SET_SYS_FROM_STRING("abiflags", ABIFLAGS);
