diff --git a/gcc/cp/g++spec.cc b/gcc/cp/g++spec.cc
index ab6482d4f5d..c9d40e7425f 100644
--- a/gcc/cp/g++spec.cc
+++ b/gcc/cp/g++spec.cc
@@ -139,6 +139,7 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
 
   /* True if we should add -shared-libgcc to the command-line.  */
   int shared_libgcc = 1;
+  int force_shared_libgcc = 0;
 
   /* The total number of arguments with the new stuff.  */
   unsigned int argc;
@@ -231,6 +232,10 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
 	  static_link = 1;
 	  break;
 
+        case OPT_shared_libgcc:
+          force_shared_libgcc = 1;
+          break;
+
 	case OPT_static_libgcc:
 	  shared_libgcc = 0;
 	  break;
@@ -305,6 +310,19 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
      libgcc.  */
 #ifndef ENABLE_SHARED_LIBGCC
   shared_libgcc = 0;
+#elif defined TARGET_PECOFF
+  /* MinGW Lite extension: fallback to static libgcc if missing shared one */
+  if (!force_shared_libgcc)
+  {
+# ifdef CROSS_DIRECTORY_STRUCTURE
+    char *target_lib = concat ("/", DEFAULT_TARGET_MACHINE, "/lib/libgcc_s.a", NULL);
+# else
+    const char *target_lib = "/lib/libgcc_s.a";
+# endif
+    char *libgcc_s = make_relative_prefix (decoded_options[0].arg, "/bin", target_lib);
+    if (libgcc_s == NULL || access (libgcc_s, F_OK) != 0)
+      shared_libgcc = 0;
+  }
 #endif
 
   /* Add one for shared_libgcc or extra static library.  */
@@ -481,6 +499,11 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
   if (shared_libgcc && !static_link)
     generate_option (OPT_shared_libgcc, NULL, 1, CL_DRIVER,
 		     &new_decoded_options[j++]);
+#ifdef TARGET_PECOFF
+  else if (!shared_libgcc && !static_link)
+    generate_option (OPT_static_libgcc, NULL, 1, CL_DRIVER,
+		     &new_decoded_options[j++]);
+#endif
 
   *in_decoded_options_count = j;
   *in_decoded_options = new_decoded_options;
