diff --git a/libgomp/config.h.in b/libgomp/config.h.in
index 46d3eac1e61..6fc0087b495 100644
--- a/libgomp/config.h.in
+++ b/libgomp/config.h.in
@@ -121,9 +121,6 @@
 /* Define to 1 if you have the <unistd.h> header file. */
 #undef HAVE_UNISTD_H
 
-/* Define to 1 if you have the `_aligned_malloc' function. */
-#undef HAVE__ALIGNED_MALLOC
-
 /* Define to 1 if you have the `__secure_getenv' function. */
 #undef HAVE___SECURE_GETENV
 
diff --git a/libstdc++-v3/config.h.in b/libstdc++-v3/config.h.in
index 0f34f65e428..ada45c98bd8 100644
--- a/libstdc++-v3/config.h.in
+++ b/libstdc++-v3/config.h.in
@@ -540,9 +540,6 @@
 /* Define to 1 if you have the `_acosl' function. */
 #undef HAVE__ACOSL
 
-/* Define to 1 if you have the `_aligned_malloc' function. */
-#undef HAVE__ALIGNED_MALLOC
-
 /* Define to 1 if you have the `_asinf' function. */
 #undef HAVE__ASINF
 
