diff --git a/gcc/configure b/gcc/configure
index 6cb58cd6ec6..b050a360134 100755
--- a/gcc/configure
+++ b/gcc/configure
@@ -12887,7 +12887,7 @@ if test "${enable_win32_utf8_manifest+set}" = set; then :
 fi
 
 
-if test "x$enable_win32_utf8_manifest" != xno; then
+if test "x$enable_win32_utf8_manifest" == xyes; then
   host_xmake_mingw=i386/x-mingw32-utf8
   host_extra_gcc_objs_mingw=utf8rc-mingw32.o
   host_extra_objs_mingw=utf8-mingw32.o
diff --git a/gcc/configure.ac b/gcc/configure.ac
index 8382b4e7b3f..1909a07e22b 100644
--- a/gcc/configure.ac
+++ b/gcc/configure.ac
@@ -1894,7 +1894,7 @@ AC_ARG_ENABLE(win32-utf8-manifest,
 AS_HELP_STRING([--enable-win32-utf8-manifest],
                [enable embedding a utf8 manifest on mingw hosts (default)])],,)
 
-if test "x$enable_win32_utf8_manifest" != xno; then
+if test "x$enable_win32_utf8_manifest" == xyes; then
   host_xmake_mingw=i386/x-mingw32-utf8
   host_extra_gcc_objs_mingw=utf8rc-mingw32.o
   host_extra_objs_mingw=utf8-mingw32.o
