diff --git a/config/mh-mingw b/config/mh-mingw
index f5fb064813f..96be2211ce6 100644
--- a/config/mh-mingw
+++ b/config/mh-mingw
@@ -1,8 +1,8 @@
 # Add -D__USE_MINGW_ACCESS to enable the built compiler to work on Windows
 # Vista (see PR33281 for details).
 BOOT_CFLAGS += -D__USE_MINGW_ACCESS -Wno-pedantic-ms-format
-CFLAGS += -D__USE_MINGW_ACCESS
-CXXFLAGS += -D__USE_MINGW_ACCESS
+override CFLAGS += -D__USE_MINGW_ACCESS
+override CXXFLAGS += -D__USE_MINGW_ACCESS
 STAGE1_CXXFLAGS += -D__USE_MINGW_ACCESS
 STAGE2_CXXFLAGS += -D__USE_MINGW_ACCESS
 STAGE3_CXXFLAGS += -D__USE_MINGW_ACCESS
