diff --git a/gcc/pretty-print.c b/gcc/pretty-print.c
index 6243aed3dac..96552db89c7 100644
--- a/gcc/pretty-print.c
+++ b/gcc/pretty-print.c
@@ -69,6 +69,7 @@ write_all (HANDLE h, const char *s, size_t n)
       the third character into *head.
    2. If the sequence begins with a character X in [0x80,0x9F], returns
       (X-0x40) and stores a pointer to the second character into *head.
+   * Case 2 is imcompatible with UTF-8, here we ignore it.
    Stores the number of ESC character(s) in *prefix_len.
    Returns 0 if no such sequence can be found.  */
 static int
@@ -93,13 +94,6 @@ find_esc_head (int *prefix_len, const char **head, const char *str)
 	  *head = r + 1;
 	  return c;
 	}
-      if (0x80 <= c && c <= 0x9F)
-	{
-	  /* Found (case 2).  */
-	  *prefix_len = 1;
-	  *head = r + 1;
-	  return c - 0x40;
-	}
       ++r;
       escaped = c == 0x1B;
     }
