diff --git a/gcc/diagnostics/sarif-sink.cc b/gcc/diagnostics/sarif-sink.cc
index 76aec2a886c..dd33315163b 100644
--- a/gcc/diagnostics/sarif-sink.cc
+++ b/gcc/diagnostics/sarif-sink.cc
@@ -55,8 +55,10 @@ along with GCC; see the file COPYING3.  If not see
 #include "backtrace.h"
 #include "xml.h"
 #include "intl.h"
+#ifdef HAVE_AF_UNIX
 #include <sys/un.h>
 #include <sys/socket.h>
+#endif
 
 namespace diagnostics {
 
@@ -4158,6 +4160,7 @@ private:
   int m_ival;
 };
 
+#ifdef HAVE_AF_UNIX
 class sarif_socket_sink : public sarif_sink
 {
 public:
@@ -4240,6 +4243,7 @@ private:
 
   unique_fd m_fd;
 };
+#endif
 
 /* Print the start of an embedded link to PP, as per 3.11.6.  */
 
@@ -4566,6 +4570,7 @@ sarif_generation_options::dump (FILE *outfile, int indent) const
   DIAGNOSTICS_DUMPING_EMIT_BOOL_FIELD (m_state_graph);
 }
 
+#ifdef HAVE_AF_UNIX
 void
 maybe_open_sarif_sink_for_socket (context &dc)
 {
@@ -4606,6 +4611,7 @@ maybe_open_sarif_sink_for_socket (context &dc)
   sink_->update_printer ();
   dc.add_sink (std::move (sink_));
 }
+#endif
 
 #if CHECKING_P
 
diff --git a/gcc/diagnostics/sarif-sink.h b/gcc/diagnostics/sarif-sink.h
index 5a9914ac571..937cc49646f 100644
--- a/gcc/diagnostics/sarif-sink.h
+++ b/gcc/diagnostics/sarif-sink.h
@@ -188,8 +188,10 @@ extern std::unique_ptr<sarif_edge>
 make_sarif_edge (const digraphs::edge &e,
 		 sarif_builder *builder);
 
+#ifdef HAVE_AF_UNIX
 extern void
 maybe_open_sarif_sink_for_socket (context &ctxt);
+#endif
 
 } // namespace diagnostics
 
diff --git a/gcc/toplev.cc b/gcc/toplev.cc
index bcf1b7604c8..c8d8e2e7842 100644
--- a/gcc/toplev.cc
+++ b/gcc/toplev.cc
@@ -2354,7 +2354,9 @@ toplev::main (int argc, char **argv)
 
   handle_common_deferred_options ();
 
+#ifdef HAVE_AF_UNIX
   diagnostics::maybe_open_sarif_sink_for_socket (*global_dc);
+#endif
 
   init_local_tick ();
 
