diff --git a/gcc/intl.c b/gcc/intl.c
index 262745f8049..d51a68223a2 100644
--- a/gcc/intl.c
+++ b/gcc/intl.c
@@ -22,6 +22,12 @@ along with GCC; see the file COPYING3.  If not see
 #include "coretypes.h"
 #include "intl.h"
 
+#ifdef _WIN32
+#include <string>
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#endif
+
 #ifdef HAVE_LANGINFO_CODESET
 #include <langinfo.h>
 #endif
@@ -55,7 +61,32 @@ gcc_init_libintl (void)
   setlocale (LC_ALL, "");
 #endif
 
+#ifdef _WIN32
+  {
+    /* Find the locale directory.
+       TODO: Use config instead of hard-coded dirs. */
+    char buf[MAX_PATH] = {0};
+    GetModuleFileNameA(NULL, buf, MAX_PATH);
+    std::string exe_path = buf;
+
+    /* \libexec\gcc\<triplet>\<version>\cc1.exe */
+    size_t pos = exe_path.rfind("\\libexec\\gcc\\");
+    if (pos == std::string::npos) {
+      /* \bin\gcc.exe */
+      pos = exe_path.rfind("\\bin\\");
+    }
+
+    if (pos == std::string::npos) {
+      (void) bindtextdomain ("gcc", LOCALEDIR);
+    } else {
+      std::string real_prefix = exe_path.substr(0, pos);
+      std::string win32_locale_dir = real_prefix + "/share/locale";
+      (void) bindtextdomain ("gcc", win32_locale_dir.c_str());
+    }
+  }
+#else
   (void) bindtextdomain ("gcc", LOCALEDIR);
+#endif
   (void) textdomain ("gcc");
 
   /* Opening quotation mark.  */
diff --git a/libcpp/init.c b/libcpp/init.c
index 3037ecd9ba6..d6b596e0073 100644
--- a/libcpp/init.c
+++ b/libcpp/init.c
@@ -26,6 +26,12 @@ along with this program; see the file COPYING3.  If not see
 #include "localedir.h"
 #include "filenames.h"
 
+#ifdef _WIN32
+#include <string>
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#endif
+
 #ifndef ENABLE_CANONICAL_SYSTEM_HEADERS
 #define ENABLE_CANONICAL_SYSTEM_HEADERS 0
 #endif
@@ -144,7 +150,32 @@ init_library (void)
       init_trigraph_map ();
 
 #ifdef ENABLE_NLS
+#ifdef _WIN32
+      {
+	/* Find the locale directory.
+	   TODO: Use config instead of hard-coded dirs. */
+	char buf[MAX_PATH] = {0};
+	GetModuleFileNameA(NULL, buf, MAX_PATH);
+	std::string exe_path = buf;
+
+	/* \libexec\gcc\<triplet>\<version>\cc1.exe */
+	size_t pos = exe_path.rfind("\\libexec\\gcc\\");
+	if (pos == std::string::npos) {
+	  /* \bin\gcc.exe */
+	  pos = exe_path.rfind("\\bin\\");
+	}
+
+	if (pos == std::string::npos) {
+	  (void) bindtextdomain (PACKAGE, LOCALEDIR);
+	} else {
+	  std::string real_prefix = exe_path.substr(0, pos);
+	  std::string win32_locale_dir = real_prefix + "/share/locale";
+	  (void) bindtextdomain (PACKAGE, win32_locale_dir.c_str());
+	}
+      }
+#else
        (void) bindtextdomain (PACKAGE, LOCALEDIR);
+#endif
 #endif
     }
 }
