diff --git a/libgfortran/caf/shmem/shared_memory.c b/libgfortran/caf/shmem/shared_memory.c
index afafd50f80b..bc822e5c787 100644
--- a/libgfortran/caf/shmem/shared_memory.c
+++ b/libgfortran/caf/shmem/shared_memory.c
@@ -40,6 +40,10 @@ see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
 #elif defined(WIN32)
 #include <windows.h>
 #include <memoryapi.h>
+#if _WIN32_WINNT < 0x0600
+__declspec(dllimport) LPVOID WINAPI
+MapViewOfFileExNuma(HANDLE, DWORD, DWORD, DWORD, SIZE_T, LPVOID, DWORD);
+#endif
 #endif
 #include <unistd.h>
 
diff --git a/libstdc++-v3/src/c++20/tzdb.cc b/libstdc++-v3/src/c++20/tzdb.cc
index b0fbfc46a6d..818ccd0145a 100644
--- a/libstdc++-v3/src/c++20/tzdb.cc
+++ b/libstdc++-v3/src/c++20/tzdb.cc
@@ -52,6 +52,10 @@
 # define WIN32_LEAN_AND_MEAN
 # include <windows.h>
 # include <psapi.h>
+# if _WIN32_WINNT < 0x0600
+extern "C" __declspec(dllimport) DWORD WINAPI
+GetDynamicTimeZoneInformation (PDYNAMIC_TIME_ZONE_INFORMATION);
+# endif
 
 # include <array>
 #endif
