diff --git a/libstdc++-v3/include/bits/ostream.h b/libstdc++-v3/include/bits/ostream.h
index d19a76ab247..961aeb1c283 100644
--- a/libstdc++-v3/include/bits/ostream.h
+++ b/libstdc++-v3/include/bits/ostream.h
@@ -112,6 +112,7 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
        *  functions in constructs like "std::cout << std::endl".  For more
        *  information, see the iomanip header.
       */
+      __attribute__((__optimize__("no-optimize-sibling-calls")))
       __ostream_type&
       operator<<(__ostream_type& (*__pf)(__ostream_type&))
       {
@@ -121,6 +122,7 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
 	return __pf(*this);
       }
 
+      __attribute__((__optimize__("no-optimize-sibling-calls")))
       __ostream_type&
       operator<<(__ios_type& (*__pf)(__ios_type&))
       {
@@ -131,6 +133,7 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
 	return *this;
       }
 
+      __attribute__((__optimize__("no-optimize-sibling-calls")))
       __ostream_type&
       operator<<(ios_base& (*__pf) (ios_base&))
       {
