diff --git a/gcc/ipa-cp.cc b/gcc/ipa-cp.cc
index 54f4e4f543c..9f61f682d58 100644
--- a/gcc/ipa-cp.cc
+++ b/gcc/ipa-cp.cc
@@ -2618,7 +2618,7 @@ propagate_bits_across_jump_function (cgraph_edge *cs, int idx,
 	 and we store it in jump function during analysis stage.  */
 
       if (!src_lats->bits_lattice.bottom_p ()
-	  || src_lats->bits_lattice.recipient_only_p ())
+	  && !src_lats->bits_lattice.recipient_only_p ())
 	{
 	  if (!op_type)
 	    op_type = ipa_get_type (caller_info, src_idx);
