diff --unified --recursive --text gcc-13.2.0.orig/gcc/gcc-main.cc gcc-13.2.0/gcc/gcc-main.cc
--- gcc-13.2.0.orig/gcc/gcc-main.cc
+++ gcc-13.2.0/gcc/gcc-main.cc
@@ -38,9 +38,43 @@
 
 extern int main (int, char **);
 
+#ifdef __MINGW32__
+
+# define WIN32_LEAN_AND_MEAN
+# include <windows.h>
+
+struct __console_cp_init_t
+{
+  UINT saved_input_cp;
+  UINT saved_output_cp;
+
+  __console_cp_init_t()
+    : saved_input_cp(GetConsoleCP())
+    , saved_output_cp(GetConsoleOutputCP())
+  {
+    UINT cp = GetACP();
+    SetConsoleCP(cp);
+    SetConsoleOutputCP(cp);
+  }
+
+  ~__console_cp_init_t()
+  {
+    SetConsoleCP(saved_input_cp);
+    SetConsoleOutputCP(saved_output_cp);
+  }
+};
+
+#else
+
+using __console_cp_init_t = int;
+
+#endif /* __MINGW32__ */
+
 int
 main (int argc, char **argv)
 {
+  __console_cp_init_t __console_cp_init;
+
   driver d (false, /* can_finalize */
 	    false); /* debug */
 
