diff --git a/gcc/main.cc b/gcc/main.cc
index 5f682c2bf52..16ce61f6dee 100644
--- a/gcc/main.cc
+++ b/gcc/main.cc
@@ -24,6 +24,11 @@ along with GCC; see the file COPYING3.  If not see
 #include "diagnostic-core.h"
 #include "toplev.h"
 
+#if defined (__MINGW32__) && defined (_UCRT)
+# include <fcntl.h>
+# include <io.h>
+#endif
+
 int main (int argc, char **argv);
 
 /* We define main() to call toplev::main(), which is defined in toplev.cc.
@@ -33,6 +38,15 @@ int main (int argc, char **argv);
 int
 main (int argc, char **argv)
 {
+#if defined (__MINGW32__) && defined (_UCRT)
+  /* UCRT randomly eats linefeeds when piping on Windows Vista. Compiling
+     any non-trivial program compiled with '-g3 -pipe' is likely to fail.
+
+     Anbang LI reported it also occasionally occurring on Windows 10. So
+     we apply the workaround without checking Windows version. */
+  _setmode (_fileno (stdout), O_BINARY);
+#endif
+
   toplev toplev (NULL, /* external_timer */
 		 true /* init_signals */);
 
