diff --git a/libstdc++-v3/src/c++17/fs_ops.cc b/libstdc++-v3/src/c++17/fs_ops.cc
index 946fefd9e44..f87f2262a12 100644
--- a/libstdc++-v3/src/c++17/fs_ops.cc
+++ b/libstdc++-v3/src/c++17/fs_ops.cc
@@ -56,6 +56,7 @@
 #ifdef _GLIBCXX_FILESYSTEM_IS_WINDOWS
 # define WIN32_LEAN_AND_MEAN
 # include <windows.h>
+# include <win32-thunk.h>
 #endif
 
 #define _GLIBCXX_BEGIN_NAMESPACE_FILESYSTEM namespace filesystem {
@@ -687,7 +688,7 @@ fs::create_hard_link(const path& to, const path& new_hard_link,
   else
     ec.clear();
 #elif defined _GLIBCXX_FILESYSTEM_IS_WINDOWS
-  if (CreateHardLinkW(new_hard_link.c_str(), to.c_str(), NULL))
+  if (Win32Thunk::CreateHardLinkW(new_hard_link.c_str(), to.c_str(), NULL))
     ec.clear();
   else
     ec = __last_system_error();
diff --git a/libstdc++-v3/src/filesystem/ops-common.h b/libstdc++-v3/src/filesystem/ops-common.h
index 1be8d754d6f..bced2ff7541 100644
--- a/libstdc++-v3/src/filesystem/ops-common.h
+++ b/libstdc++-v3/src/filesystem/ops-common.h
@@ -44,6 +44,7 @@
 
 #ifdef _GLIBCXX_FILESYSTEM_IS_WINDOWS
 # include <wchar.h>
+# include <win32-thunk.h>
 #endif
 
 #ifdef NEED_DO_COPY_FILE
@@ -121,7 +122,7 @@ namespace __gnu_posix
   using stat_type = struct ::__stat64;
 
   inline int stat(const wchar_t* path, stat_type* buffer)
-  { return ::_wstat64(path, buffer); }
+  { return Win32Thunk::_wstat64(path, buffer); }
 
   inline int lstat(const wchar_t* path, stat_type* buffer)
   {
