diff --git a/bfd/bfdio.c b/bfd/bfdio.c
index ab3e28ea204..4add80f62d6 100644
--- a/bfd/bfdio.c
+++ b/bfd/bfdio.c
@@ -117,6 +117,11 @@ _bfd_real_fopen (const char *filename, const char *modes)
     }
 
 #elif defined (_WIN32)
+  OSVERSIONINFOA osvi = {sizeof (OSVERSIONINFOA)};
+  GetVersionExA (&osvi);
+  if (osvi.dwPlatformId < VER_PLATFORM_WIN32_NT)
+    return close_on_exec (fopen (filename, modes));
+
   /* PR 25713: Handle extra long path names possibly containing '..' and '.'.  */
    wchar_t **     lpFilePart = {NULL};
    const wchar_t  prefix[] = L"\\\\?\\";
