diff --unified --recursive --text gettext-0.22.5.orig/gettext-runtime/intl/gnulib-lib/windows-mutex.c gettext-0.22.5/gettext-runtime/intl/gnulib-lib/windows-mutex.c
--- gettext-0.22.5.orig/gettext-runtime/intl/gnulib-lib/windows-mutex.c
+++ gettext-0.22.5/gettext-runtime/intl/gnulib-lib/windows-mutex.c
@@ -21,6 +21,7 @@
 
 /* Specification.  */
 #include "windows-mutex.h"
+#include "win32-thunk.h"
 
 #include <errno.h>
 
diff --unified --recursive --text gettext-0.22.5.orig/gettext-runtime/intl/gnulib-lib/windows-recmutex.c gettext-0.22.5/gettext-runtime/intl/gnulib-lib/windows-recmutex.c
--- gettext-0.22.5.orig/gettext-runtime/intl/gnulib-lib/windows-recmutex.c
+++ gettext-0.22.5/gettext-runtime/intl/gnulib-lib/windows-recmutex.c
@@ -21,6 +21,7 @@
 
 /* Specification.  */
 #include "windows-recmutex.h"
+#include "win32-thunk.h"
 
 #include <errno.h>
 
